### ISSUE TITLE
Validation ordering can be counter-intuitive

### ISSUE THREAD

**Issue by dannybster** [nestjs/nest#9269]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behavior

When utilising `ValidationPipe` and the associated `Body` decorator, with a custom decorator that throws an Exception the behaviour might be considered counterintuitive e.g.

With the Controller:
```ts
import { Body, Controller, Post } from '@nestjs/common';
import { Name } from './decorators/name.decorator';
import { GreetingDto } from './dtos/greeting.dto';

@Controller()
export class AppController {
  @Post('/greetings')
  postGreeting(@Name() name: string, @Body() greeting: GreetingDto): string {
    return `${greeting.greeting} ${name}, how are you today?`;
  }
}
```

That utilises the "built-in" `Body` decorator and the "custom" decorator `Name` which is defined as per below:
```ts
import {
  createParamDecorator,
  ExecutionContext,
  NotFoundException,
} from '@nestjs/common';
import { Request } from 'express';

export const Name = createParamDecorator(
  (_data: any, context: ExecutionContext): string => {
    const http = context.switchToHttp();
    const request = http.getRequest<Request>();
    const { name } = request.body;
    if (name === 'Not Dan') {
      throw new NotFoundException(
        `The name ${name} could not be found. Are you sure this person is real?`,
      );
    }
    return name;
  },
);
```

I would expect the following tests to pass:
```ts
import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import { getReasonPhrase, StatusCodes } from 'http-status-codes';
import * as request from 'supertest';
import { AppModule } from './../src/app.module';

const createdBody = { greeting: 'Hey', name: 'Dan' };
const notFoundBody = { greeting: 'Hey', name: 'Not Dan' };
// This body is applicable to two errors:
// - A not found error as it has the name Not Dan
// - A bad request error as it is missing the property greeting
// I would expect a bad request error as I expected validation
// to take a higher (or equal) priority to customer parameter
// decorators.
const badRequestBody = { name: 'Not Dan' };

const { BAD_REQUEST, CREATED, NOT_FOUND } = StatusCodes;

describe('AppController (e2e)', () => {
  let app: INestApplication;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  describe('When a POST request is made to /greetings', () => {
    describe(`And the request body is ${JSON.stringify(createdBody)}`, () => {
      it(`Then it should respond with an HTTP ${CREATED} and a greeting.`, () => {
        return request(app.getHttpServer())
          .post('/greetings')
          .expect(CREATED)
          .expect(
            `${createdBody.greeting} ${createdBody.name}, how are you today?`,
          )
          .send(createdBody);
      });
    });

    describe(`And the request body is ${JSON.stringify(notFoundBody)}`, () => {
      it(`Then it should respond with an HTTP ${NOT_FOUND}..`, () => {
        return request(app.getHttpServer())
          .post('/greetings')
          .expect(NOT_FOUND)
          .expect({
            statusCode: NOT_FOUND,
            message: `The name ${notFoundBody.name} could not be found. Are you sure this person is real?`,
            error: getReasonPhrase(NOT_FOUND),
          })
          .send(notFoundBody);
      });
    });

    describe(`And the request body is ${JSON.stringify(
      badRequestBody,
    )}`, () => {
      it.only(`Then it should respond with an HTTP ${BAD_REQUEST}.`, () => {
        return request(app.getHttpServer())
          .post('/greetings')
          .expect(BAD_REQUEST)
          .send(badRequestBody);
      });
    });
  });
});
```
That is:
- With a valid request body everything should work fine and the response should be 201
- With a valid request body but with a name property referencing a resource that does not exist (name === 'Not Dan') I would expect a 404
- With an invalid request body AND and said body references a resource that does not exist (name === 'Not Dan') I would expect a 400

From debugging I can see that the custom `Name` decorator runs first (regardless of parameter ordering in the controller method signature) and throws the `NotFoundException` and then the ValidationPipe runs for the `Body`. I'm not sure this is intended behaviour, although it appears it might be as per the ordering declared in [The NestJS Lifecycle](https://docs.nestjs.com/faq/request-lifecycle ) i.e. I think I see the "custom" decorator `Name` running in the context of an interceptor based on debugging, but it does feel unintuitive that custom parameter decorators (running as interceptors?)  run before the Validation Pipe.

NOTE: I refer to custom parameter decorators running as interceptors due to the stack trace when they run referencing interceptors/interceptors-consumer.js:
```
console.log
    IN DECORATOR Error:
        at /Users/Dan/Dropbox/shipd/scratchpad/src/decorators/name.decorator.ts:10:33
        at /Users/Dan/Dropbox/shipd/scratchpad/node_modules/@nestjs/core/helpers/context-utils.js:40:28
        at resolveParamValue (/Users/Dan/Dropbox/shipd/scratchpad/node_modules/@nestjs/core/router/router-execution-context.js:145:31)
        at Array.map (<anonymous>)
        at pipesFn (/Users/Dan/Dropbox/shipd/scratchpad/node_modules/@nestjs/core/router/router-execution-context.js:150:45)
        at /Users/Dan/Dropbox/shipd/scratchpad/node_modules/@nestjs/core/router/router-execution-context.js:37:36
        at InterceptorsConsumer.intercept (/Users/Dan/Dropbox/shipd/scratchpad/node_modules/@nestjs/core/interceptors/interceptors-consumer.js:11:20)
        at /Users/Dan/Dropbox/shipd/scratchpad/node_modules/@nestjs/core/router/router-execution-context.js:46:60
        at /Users/Dan/Dropbox/shipd/scratchpad/node_modules/@nestjs/core/router/router-proxy.js:9:23
        at Layer.handle [as handle_request] (/Users/Dan/Dropbox/shipd/scratchpad/node_modules/express/lib/router/layer.js:95:5)
```

In case this is confusing I have attached a sample repo that I hope helps, but please feel free to ask anything further.

### Minimum reproduction code

https://github.com/dannybster/DecoratorOrdering.git

### Steps to reproduce

1. `npm ci`
2. `npm run test:e2e`
3. See console output and examine app.e2e-spec.ts

### Expected behavior

What I would expect to see from a purely intuitive standpoint, and hence this might be something that needs highlighting in the docs, is that `Body` and the associated validation, and the customer decorator `Name` appear to have the same precedence i.e. the parameter decorator cannot run before request validation.

I think what is happening is that the Validation Pipe runs after the decorators and so errors thrown in the decorators take precedence but I could be wrong.

### Package

- [ ] I don't know. Or some 3rd-party package
- [X] <code>@nestjs/common</code>
- [X] <code>@nestjs/core</code>
- [ ] <code>@nestjs/microservices</code>
- [ ] <code>@nestjs/platform-express</code>
- [ ] <code>@nestjs/platform-fastify</code>
- [ ] <code>@nestjs/platform-socket.io</code>
- [ ] <code>@nestjs/platform-ws</code>
- [ ] <code>@nestjs/testing</code>
- [ ] <code>@nestjs/websockets</code>
- [ ] Other (see below)

### NestJS version

81.1.

### Packages versions

```json
Nest CLI Version : 8.2.1
platform-express : 8.3.1
schematics : 8.0.7
testing : 8.3.1
common : 8.3.1
core : 8.3.1
cli : 8.2.1
```

### Node.js version

14.17.5

### In which operating systems have you tested?

- [X] macOS
- [ ] Windows
- [ ] Linux

**Comment by jmcdo29** [comment]
I believe this ends up making sense, when thinking about how decorators work and what Nest is doing under the hood. So we have to consider the different kinds of decorators at play here (method and parameter) and what Nest is really doing for us with how all of this is set up, right? So lets start with the method decorators, `@UsePipes()`, `@UseInterceptor()`, etc. I know you're not using them here, just the global `APP_PIPE`, but it's still good to think about

### Method Decorators

What these decorators (`@UsePipes()`, `@UseInterceptors()`, etc) are doing is setting up metadata to be read later about what pipes and interceptors to run when this route handler is hit. Nest will read that metadata when a route is hit and end up calling these enhancers for us, kind of like middleware. Interceptors run across the entire route handler, in true middleware fashion. Pipes run against every parameter of the route handler, so keep that in mind

### Parrameter Decorators

These decorators (`@Body()`, `@Param()`, your custom `@Name()`) also end up setting up metadata for Nest to read later. Each of these parameter decorators Nest has a special method associated with it, for how to actually get the data from the request object to the proper position. When we apply a callback to `createParamDecorator` that's what we're doing, telling Nest where to find this custom data. The `@Body()` is simply `(data?: string) => data ? req.body[data] : req.body`. Yours here ends up having the possibility to throw an error which we'll want to keep in mind.

### Execution Order

So before we get to executing any pipes, we need to make sure that all of our method parameters are resolved and correct. For `@Body()`/`@Param()`/etc this is pretty simple, just get the related `req.property` and boom, we're good. With this custom `@Name()` you've created Nest calls the callback for the decorator with the `context` object and calls through your code. In your test case, you end up getting an error from the `name === 'Not Dan'` line. In the current execution order, __all__ parameter decorators need to resolve to values before calling to pipes. I suppose we could switch it to where we resolve a param decorator at a time and send it through the appropriate pipes, but that's not currently how its set up. 

### Closing Thoughts

Generally, I'd probably suggest moving the check in the decorator to a dedicated pipe. It'll lead to less headache down the road

**Comment by dannybster** [comment]
@jmcdo29 this is massively helpful, thanks for taking the time to explain this in such depth. 

I was definitely missing some understanding of the workflow you outline above (which makes total sense now), which led me to write the software the way I did i.e. I did not realise that all the method parameters were resolved first (so the data was available) then the pipes ran. As such I will certainly move the check out into a dedicated Pipe. In fact, the [Lifecycle](https://docs.nestjs.com/faq/request-lifecycle) definitely confused me further making it seem much more top-down leading me to believe the parameter decorators all ran last.

I've been over the docs a ton and didn't find the information you outline (could have been looking in the wrong place), is this in the docs and if not is there any chance your explanation could be added so that others don't make the same mistake I did. I do think it would make a fantastic addition to the Pipes docs.

**Comment by jmcdo29** [comment]
We don't talk in the docs much about the resolution order of the decorators themselves because it's usually such an advanced topic, though I think adding an `Advanced` section eventually is something I'd like to do that helps get into the nitty-gritty of TS decorators and what Nest is really doing with them (at least mostly). 

I tried to make it clear in the lifecycle docs that it was the parameter _pipes_ that were executed last,m before the route handler itself, but I see that I may need to add more clarification around that.

**Comment by dannybster** [comment]
Make sense. I think given that I am new to NestJs it is likely that I do not understand the fundamentals, partly because the docs do such a good job of getting a developer up and running without having to understand much in-depth.

I probably need to understand pipes further, are there any other docs you suggest I should read or should I just be diving into the source?

**Comment by jmcdo29** [comment]
It seems like what you're looking to learn more on, rather than just how the pipes work, is an in depth knowledge of how Nest works. We don't have any docs on that, it's kind of the abstracted part from express and fastify. You can definitely look through the source code if you want, or you can just read the [pipes page](https://docs.nestjs.com/pipes) and go from there. I will say, the source code takes some time to wrap your head around with what all is happening. It makes sense, there's just a lot to it.

**Comment by dannybster** [comment]
Excellent, with the explanation you've provided and re-reading the Pipes section the following now makes total sense:

"In both cases, pipes operate on the arguments being processed by a [controller route handler](https://docs.nestjs.com/controllers#route-parameters). Nest interposes a pipe just before a method is invoked, and the pipe receives the arguments destined for the method and operates on them. Any transformation or validation operation takes place at that time, after which the route handler is invoked with any (potentially) transformed arguments."

Agree, I'm definitely looking to gather in-depth knowledge of NestJs, it's going to be the backbone of a multi-year endeavour, white-label product. Regarding source and wrapping my head around it, any hints and tips e.g. where to start, concepts I should probably pick up?

Thanks for your help, it's massively appreciated. If I can ever repay the favour please let me know!

**Comment by dannybster** [comment]
Hey @jmcdo29, happy for you to close this out unless you want to keep it around as a reminder to update docs.

Thanks for all your help!!

**Comment by YairLevi** [comment]
Hey, I hope that even though this is closed I'll still get some help regarding my problem. 
I'm new to NestJS and also came across a similar problem (hence, how I got here).

I have a route that accepts a body of the following type:
```typescript
export class CreateUserRequest {
  authId: string
}
```
I created a pipe that transforms this data into the type of:
```typescript
import { IsEmail, IsPhoneNumber, IsDefined } from "class-validator";

export class CreateUserDto {

  @IsDefined()
  authId: string

  @IsDefined()
  name: string

  @IsDefined()
  @IsEmail()
  email: string

  phone: string

  photo: string

  passwordHash: string
}
```
The pipe's signature is:
```typescript
export class CreateUserPipe implements PipeTransform<CreateUserRequest, Promise<CreateUserDto>>
```

When I specify the type of the `@Body()` in the handler method, If i specify it using `CreateUserDto`:
```typescript
@Post('create')
async create(@Body(CreateUserPipe) userDto: CreateUserDto) {
  await this.userService.create(userDto)
}
```
It throws an error, because (from what I understand, and correct me if I'm wrong) the decorator of `@Body()` first is executed checking if the body matches the type of `CreateUserDto`, while I thought that first the pipes were supposed to run, and then this check would occur.
However, if I switch to the type of body that comes straight from the request:
```typescript
@Post('create')
// ----------------------------------------  right under here ----
async create(@Body(CreateUserPipe) userDto: CreateUserRequest) {
  await this.userService.create(userDto)
}
```
Then it throws me an error, because the handler thinks `userDto` is of type `CreateUserRequest`, while it has transformed using the pipe to type `CreateUserDto` (the handler needs it to be of type `CreateUserDto`).
How can I solve this? I know one solution that worked for me was just to not specify the type of `userDto`, but is there some other solution where I do specify the type for more clarity in code?
And in general, is there a way to specify the type that an argument should be at the end of a pipe-line? (pun intended)

Any help and support + more insight as to how to think about pipes would be much appreciated!
